README.md
RubberStamp :mailbox:
================

![Feature Image](images/FeatureImage.png)

RubberStamp is an Android library that makes it easy for you to watermark your images.

Screenshots
------------
![Screenshots](images/screenshots.png)


Features
---------

* Add a watermark to your images.
* It can be text or another image.
* Multiple ways and features to customize how the watermark looks including attributes like font, color, opacity, size, shadow properties, etc.
* Flexible API that has multiple pre-defined options to get started quickly.
* Ability to tile your watermark across an image.


Setup
------
The library is pushed to Maven Central as an AAR, so you just need to add the following to your build.gradle file:
```java
dependencies {
    compile ‘com.vinaygaba:rubberstamp:1.0.0’
}
```

Usage
------
Using RubberStamp is extremely easy.

First, define the characteristics of your watermark using RubberStampConfig

```java
RubberStampConfig config = new RubberStampConfigBuilder()
              .base(R.drawable.lenna)
              .rubberStamp("Watermark")
              .alpha(100)
              .textColor(Color.BLACK)
              .textBackgroundColor(Color.WHITE)
              .textShadow(0.1f,  5, 5, Color.BLUE)
              .textSize(90)
              .rotation(-45)
              .rubberStampPosition(RubberStamp.CENTER)
              .build();
```

That's all that is needed really. All you need to do next is just pass this config to the addStamp method of the RubberStamp class and voila!
```java
RubberStamp rubberStamp = new RubberStamp(this);
rubberStamp.addStamp(config);
```

Attribute Usage & Documentation
--------------------------------
All the attributes listed below are part of RubberStampConfig.

##### I. `base`
Use this to set the base image on top of which the watermark will be drawn. This can be a bitmap or a drawable.

```java
config.base(bitmap);

config.base(R.id.image);
```

##### II. `rubberStamp`
The rubberstamp is the actual watermark that will be drawn on top of the base image. This can either be a bitmap or a string.

```java
// String watermark
config.rubberStamp("Watermark" );
```
![Screenshots](images/rubberstamp1.png)

```java
// Bitmap watermarkBitmap
config.rubberStamp(bitmap);
```
![Screenshots](images/rubberstamp2.png)

If you want to use a drawable, convert it to a bitmap and use that. You would do that using:

```java
Bitmap bitmap = BitmapFactory.decodeResources(getResources(), R.drawable.logo);

```

##### III. `rubberStampPosition`
The library provides 9 pre defined positions to display the location of the rubberstamp. They are as follows:

TOP_LEFT<br/>
TOP_CENTER<br/>
TOP_RIGHT<br/>
BOTTOM_LEFT<br/>
BOTTOM_CENTER<br/>
BOTTOM_RIGHT<br/>
CENTER_LEFT<br/>
CENTER<br/>
CENTER_RIGHT<br/>

```java
// RubberStampPosition position
config.rubberStampPosition(RubberStampPosition.BOTTOM_RIGHT);

```
![Screenshots](images/rubberStampPosition.png)

In additon, if you would like to specify the exact position of the rubberstamp, you can pass the the RubberStampPosition to be `CUSTOM` and use the following constructor to specify the position.

```java
// RubberStampPosition position, int xCoordinate, int yCoordinate
config.rubberStampPosition(RubberStampPosition.CUSTOM, 100, 100);
```

There is another special position, `TILE` that you can use in order to tile the rubberstamp across the base image.
This is a fairly common use case for watermarking software so it made sense for the library
to support it. You can use it in the following way:

```java
// RubberStampPosition position
config.rubberStampPosition(RubberStampPosition.TILE);
```
![Screenshots](images/tiling.png)

##### IV. `alpha`

Use alpha to change the opacity of your rubberstamp. It accepts an int value between 0 and 255.

```java
//int alpha
config.alpha(255);
```
![Screenshots](images/textalpha.png)

##### V. `rotation`

Rotation does exactly what you'd imagine it to: it rotates your rubberstamp. It expects a float value
to be passed to it.

```java
config.rotation(45);
config.rotation(-45);
```
![Screenshots](images/rotation.png)

##### VI. `margin`
Margin lets you specify the x & y offset for your rubberstamp after you have selected its RubberStampPosition.
This is to give the user the ability to position his rubberstamp at a more precise location if the defaults RubberStampPosition presets are not good enough.

```java
// int xMargin, int yMargin
config.margin(-30, 30);
```
### Text RubberStamp specific attributes

There are some additional attributes that this library provides when you pass a String as the RubberStamp.
These attributes won't have any side effects when you pass a bitmap as a rubberstamp and will be disregarded.
Most of them are pretty self explanatory.

##### VII. `textColor`

Sets the text color of the rubberstamp.
```java
config.textColor(Color.parseColor("#FFB6C1"));
```
![Screenshots](images/textColor.png)

##### VIII. `textBackgroundColor`

This lets you specify a background color for your text rubberstamp.

<b>Note</b> This attribute does not work when the position is set to RubberStampPosition.TILE

```java
//int color
config.textBackgroundColor(Color.WHITE);
```
![Screenshots](images/textBackground.png)

##### IX. `textSize`
Sets the size of the text rubberstamp.

```java
//int size
config.textSize(40);
```
![Screenshots](images/textSize.png)

##### X. `textShader`

This lets you specify a custom shader that you can use to customize the watermark.
Honestly, the sky is the limit when it comes to Shaders. Here is an example shader that
paints my watermark in rainbow colors.

```java
RubberStamp rubberStamp = new RubberStamp(this);
int[] rainbow = {Color.RED, Color.YELLOW, Color.GREEN, Color.BLUE, Color.MAGENTA};
Shader shader = new LinearGradient(0, 0, 0, logo.getWidth(), rainbow,
        null, Shader.TileMode.MIRROR);
Matrix matrix = new Matrix();
matrix.setRotate(90);
shader.setLocalMatrix(matrix);

//set the shader
config.textShader(shader);
```
![Screenshots](images/textShader.png)

##### XI. `textShadow`
This lets you specify a shadow for your text rubberstamp.
<b>Note:</b> No shadow will be displayed if the blur radius is set to 0. This is how the paint
API in Android behaves underneath as well.
```java
//float blurRadius, float xOffset, float yOffset, int color
config.textShadow(1.0f, 5, 5, Color.BLUE);
```
![Screenshots](images/textShadow.png)

##### XII. `textFont`
Use this to specify a custom font for your text rubberstamp.

```java
//String fontpath
config.textFont("fonts/champagne.ttf");
```
Credits
-----------------
Author: Vinay Gaba (vinaygaba@gmail.com)

<a href="https://plus.google.com/+Vinaygaba">
  <img alt="Follow me on Google+"
       src="https://github.com/gabrielemariotti/cardslib/raw/master/demo/images/g+64.png" />
</a>
<a href="https://twitter.com/vinaygaba">
  <img alt="Follow me on Twitter"
       src="https://github.com/gabrielemariotti/cardslib/raw/master/demo/images/twitter64.png" />
</a>
<a href="https://www.linkedin.com/in/vinaygaba">
  <img alt="Follow me on LinkedIn"
       src="https://github.com/gabrielemariotti/cardslib/raw/master/demo/images/linkedin.png" />
</a>


License
-------

    Copyright 2017 Vinay Gaba

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
Version 1.0.0
--------------

Initial release

sample/src/main/java/com/vinaygaba/sample/MainActivity.java
package com.vinaygaba.sample;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.LinearGradient;
import android.graphics.Matrix;
import android.graphics.Shader;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RadioGroup;
import android.widget.SeekBar;
import android.widget.Spinner;
import android.widget.Switch;
import android.widget.TextView;

import com.pes.androidmaterialcolorpickerdialog.ColorPicker;
import com.pes.androidmaterialcolorpickerdialog.ColorPickerCallback;
import com.vinaygaba.rubberstamp.RubberStamp;
import com.vinaygaba.rubberstamp.RubberStampConfig;
import com.vinaygaba.rubberstamp.RubberStampPosition;

import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.functions.Func0;
import rx.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {

    public static final String TAG = "Sample";
    private ImageView mImageView;
    private RadioGroup mRadioGroup;
    private Bitmap mBaseBitmap;
    private Button mGenerateButton;
    private SeekBar mAlphaSeekBar, mRotationSeekBar, mTextSizeSeekBar;
    private Spinner mRubberStampPosition, mTextFonts;
    private LinearLayout mTextLayoutWrapper;
    private TextView mTextColor, mTextBackgroundColor;
    private EditText mRubberStampText;
    private ColorPicker mColorPicker;
    private RubberStamp mRubberStamp;
    @ColorInt private int mTextColorValue = Color.RED;
    @ColorInt private int mTextBackgroundColorValue = Color.TRANSPARENT;
    private Switch mShaderSwitch;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        init();
        setListeners();
        setDefaults();
    }

    public void init() {
        mRadioGroup = (RadioGroup) findViewById(R.id.radioGroup);
        mImageView = (ImageView) findViewById(R.id.imageView);
        mBaseBitmap = BitmapFactory.decodeResource(getResources(),
                R.drawable.lenna);
        mGenerateButton = (Button) findViewById(R.id.generateRubberStamp);
        mAlphaSeekBar = (SeekBar) findViewById(R.id.alphaSeekBar);
        mRotationSeekBar = (SeekBar) findViewById(R.id.rotationSeekBar);
        mTextSizeSeekBar = (SeekBar) findViewById(R.id.textSizeSeekBar);
        mRubberStampPosition = (Spinner) findViewById(R.id.rubberStampPositions);
        mTextFonts = (Spinner) findViewById(R.id.textFonts);
        mTextLayoutWrapper = (LinearLayout) findViewById(R.id.textLayoutWrapper);
        mTextColor = (TextView) findViewById(R.id.textColor);
        mTextBackgroundColor = (TextView) findViewById(R.id.textBackgroundColor);
        mRubberStampText = (EditText) findViewById(R.id.rubberStampEditText);
        mShaderSwitch = (Switch) findViewById(R.id.shaderSwitch);
        mColorPicker = new ColorPicker(this, 255, 0, 0);
        mRubberStamp = new RubberStamp(this);
    }

    public void setListeners() {
        mRadioGroup.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(RadioGroup group, int checkedId) {
                if(checkedId == R.id.textRubberStamp) {
                    mTextLayoutWrapper.setVisibility(View.VISIBLE);
                } else {
                    mTextLayoutWrapper.setVisibility(View.GONE);
                }
            }
        });

        mGenerateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                generateRubberStamp();
            }
        });

        mTextColor.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                mColorPicker.show();
                mColorPicker.setCallback(new ColorPickerCallback() {
                    @Override
                    public void onColorChosen(@ColorInt int color) {
                        Drawable background = mTextColor.getCompoundDrawables()[2];
                        GradientDrawable gradientDrawable = (GradientDrawable) background;
                        gradientDrawable.setColor(color);
                        mTextColorValue = color;
                        mColorPicker.dismiss();
                    }
                });
                return false;
            }
        });

        mTextBackgroundColor.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                mColorPicker.show();
                mColorPicker.setCallback(new ColorPickerCallback() {
                    @Override
                    public void onColorChosen(@ColorInt int color) {
                        Drawable background = mTextBackgroundColor.getCompoundDrawables()[2];
                        GradientDrawable gradientDrawable = (GradientDrawable) background;
                        gradientDrawable.setColor(color);
                        mTextBackgroundColorValue = color;
                        mColorPicker.dismiss();
                    }
                });
                return false;
            }
        });
    }

    public void setDefaults() {
        Drawable background = mTextBackgroundColor.getCompoundDrawables()[2];
        GradientDrawable gradientDrawable = (GradientDrawable) background;
        gradientDrawable.setColor(mTextBackgroundColorValue);
    }

    public void generateRubberStamp() {
        RubberStampConfig config;
        int alpha = mAlphaSeekBar.getProgress();
        int rotation = mRotationSeekBar.getProgress();
        RubberStampPosition rubberStampPosition =
                convertToRubberStampPosition(mRubberStampPosition.getSelectedItemPosition());

        if(mRadioGroup.getCheckedRadioButtonId() == R.id.bitmapRubberStamp) {
            Bitmap logo = BitmapFactory.decodeResource(getResources(),
                    R.drawable.logo);

            config = new RubberStampConfig.RubberStampConfigBuilder()
                    .base(mBaseBitmap)
                    .rubberStamp(logo)
                    .alpha(alpha)
                    .rotation(rotation)
                    .rubberStampPosition(rubberStampPosition)
                    .build();
        } else {
            String path = convertFontPositionToPath(mTextFonts.getSelectedItemPosition());
            Shader shader = getShader();
            config = new RubberStampConfig.RubberStampConfigBuilder()
                    .base(mBaseBitmap)
                    .alpha(alpha)
                    .rotation(rotation)
                    .rubberStampPosition(rubberStampPosition)
                    .rubberStamp(mRubberStampText.getText().toString())
                    .textColor(mTextColorValue)
                    .textBackgroundColor(mTextBackgroundColorValue)
                    .textShader(shader)
                    .textFont(path)
                    .textSize(mTextSizeSeekBar.getProgress())
                    .build();
        }


        Observable<Bitmap> observable = getBitmap(mRubberStamp, config);
        observable.subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .single()
                .subscribe(new Action1<Bitmap>() {
                    @Override
                    public void call(Bitmap bitmap) {
                        mImageView.setImageBitmap(bitmap);
                    }
                }, new Action1<Throwable>() {
                    @Override
                    public void call(Throwable throwable) {
                        Log.e(TAG, throwable.getMessage());
                    }
                });
    }

    private RubberStampPosition convertToRubberStampPosition(int selectedItemPosition) {
        switch (selectedItemPosition) {
            case 0: return RubberStampPosition.TOP_LEFT;

            case 1: return RubberStampPosition.TOP_CENTER;

            case 2: return RubberStampPosition.TOP_RIGHT;

            case 3: return RubberStampPosition.CENTER_LEFT;

            case 4: return RubberStampPosition.CENTER;

            case 5: return RubberStampPosition.CENTER_RIGHT;

            case 6: return RubberStampPosition.BOTTOM_LEFT;

            case 7: return RubberStampPosition.BOTTOM_CENTER;

            case 8: return RubberStampPosition.BOTTOM_RIGHT;

            case 9: return RubberStampPosition.CUSTOM;

            case 10: return RubberStampPosition.TILE;

            default: return RubberStampPosition.CENTER;
        }
    }

    public String convertFontPositionToPath(int position) {
        String basePath = "fonts/";
        switch (position) {
            case 0: return basePath + "bebasneue.otf";
            case 1: return basePath + "caviardreams.ttf";
            case 2: return basePath + "champagne.ttf";
            default: return basePath + "bebasneue.otf";
        }
    }

    public Observable<Bitmap> getBitmap(final RubberStamp rubberStamp,
                                        final RubberStampConfig config) {
        return Observable.defer(new Func0<Observable<Bitmap>>() {
            @Override
            public Observable<Bitmap> call() {
                return Observable.just(rubberStamp.addStamp(config));
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    public Shader getShader() {
        if (mShaderSwitch.isChecked()) {
            int[] rainbow = {Color.RED, Color.YELLOW, Color.GREEN, Color.BLUE, Color.MAGENTA};
            Shader shader = new LinearGradient(0, 0, 0, 100, rainbow,
                    null, Shader.TileMode.MIRROR);

            Matrix matrix = new Matrix();
            matrix.setRotate(90);
            shader.setLocalMatrix(matrix);

            return shader;
        }
        return null;
    }
}